notebooks/sequence_and_schema.ipynb

```diff
@@ -0,0 +1,172 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Data sequence and schema diagrams\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Data Sequences\n",
+    "### File apis\n",
+    "\n",
+    "```mermaid\n",
+    "---\n",
+    "title: Data sequence - POST /upload\n",
+    "---\n",
+    "\n",
+    "sequenceDiagram\n",
+    "    Django->>S3: file key, content\n",
+    "    Django->>Core: file key\n",
+    "    Core->>Workers: file key\n",
+    "    Core->>Elastic: file key\n",
+    "    S3->>Workers: file content\n",
+    "    Workers->>Elastic: chunk key, content\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Chat APIs\n",
+    "\n",
+    "\n",
+    "```mermaid\n",
+    "---\n",
+    "title: Data sequence - POST /chat/vanilla\n",
+    "---\n",
+    "\n",
+    "sequenceDiagram\n",
+    "    Django->>Core: ChatHistory.messages[]\n",
+    "    Core->>LLM API: ChatHistory.messages[]\n",
+    "\n",
+    "```\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "```mermaid\n",
+    "---\n",
+    "title: Data sequence - POST /chat/rag\n",
+    "---\n",
+    "\n",
+    "sequenceDiagram\n",
+    "    Django->> Core: ChatHistory.messages[], File[].uuid\n",
+    "    Core->>Embedding API: ChatHistory.messages[]\n",
+    "    Embedding API->>Core: ChatHistory.messages[].embeddings\n",
+    "    Elastic->>Core: File[].Chunk[].embeddings\n",
+    "    Core->>LLM API: ChatHistory.messages[].embeddings + File[].Chunk[].embeddings\n",
+    "\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Schema\n",
+    "\n",
+    "```mermaid\n",
+    "---\n",
+    "title: Django schema\n",
+    "---\n",
+    "\n",
+    "erDiagram\n",
+    "    User }|--|{ UserGroup : \"UserGroup.users\"\n",
+    "    User {\n",
+    "        UUID uuid\n",
+    "        string name\n",
+    "    }\n",
+    "    UserGroup {\n",
+    "        UUID uuid\n",
+    "        string name\n",
+    "        UUID[] users\n",
+    "    }\n",
+    "    \n",
+    "    FileRecord }|--|| UserGroup: \"FileRecord.owner\"\n",
+    "    FileRecord {\n",
+    "        UUID uuid\n",
+    "        UUID owner\n",
+    "        string key \n",
+    "    }\n",
+    "\n",
+    "    UserGroup ||--|{ Message : \"Mesage.owner\"\n",
```

Comment:
Do we need Messages to have a concept of an owner, if they are part of a Chat History which also has that owner?

Reply:
it seems reasonable that a message would belong to chat history?

Reply:
it seems reasonable that a message would belong to chat history?

maybe we could call "Message" -> "ChatMessage"?

Reply:
Agree on the `ChatMessage` notation to avoid conflation with Alert messages that we might implement in future. A ChatMessage will always belong to a `ChatHistory`. However, it might make usage metrics easier for individual messages to have ownership by user directly in object. Equally, having individual ownership of `ChatMessages` enables us to explore group chats in future if we need it

Reply:
agree with both points. lets avoid polymorphism in the database at all costs. for UserGroups lets subclass django's Group model (which i think has a many-to-many relationship to the User)?